### ISSUE TITLE
Seems hyperref v6.86b breaks hyperlink-creation with amsthm v2.20.4

### ISSUE THREAD

**Issue by Ulrich-Diez** [ho-tex/hyperref#66]
Seems **hyperref v6.86b** breaks hyperlink-creation with **amsthm v2.20.4**.

More details at:

[TeX LaTeX StackExchange - Inaccurate hyperlink to customly defined theorems](https://tex.stackexchange.com/a/442917/118714)

The definition of `\@thm` in **amsthm** is:

    \def\@thm#1#2#3{%
      \ifhmode\unskip\unskip\par\fi
      \normalfont
      \trivlist
      \let\thmheadnl\relax
      \let\thm@swap\@gobble
      \thm@notefont{\fontseries\mddefault\upshape}%
      \thm@headpunct{.}% add period after heading
      \thm@headsep 5\p@ plus\p@ minus\p@\relax
      \thm@space@setup
      #1% style overrides
      \@topsep \thm@preskip               % used by thm head
      \@topsepadd \thm@postskip           % used by \@endparenv
      \def\@tempa{#2}\ifx\@empty\@tempa
        \def\@tempa{\@oparg{\@begintheorem{#3}{}}[]}%
      \else
        \refstepcounter{#2}%
        \def\@tempa{\@oparg{\@begintheorem{#3}{\csname the#2\endcsname}}[]}%
      \fi
      \@tempa
    }

Seems **hyperref** does override that definition by

    \def\@thm#1#2#3{%
      \ifhmode
        \unskip\unskip\par
      \fi
      \normalfont
      \trivlist
      \let\thmheadnl\relax
      \let\thm@swap\@gobble
      \thm@notefont{\fontseries\mddefault\upshape}%
      \thm@headpunct{.}% add period after heading
      \thm@headsep 5\p@ plus\p@ minus\p@\relax
      \thm@space@setup
      #1% style overrides
      \@topsep \thm@preskip               % used by thm head
      \@topsepadd \thm@postskip           % used by \@endparenv
      \def\dth@counter{#2}%
      \ifx\@empty\dth@counter
        \def\@tempa{%
          \@oparg{\@begintheorem{#3}{}}[]%
        }%
      \else
        \H@refstepcounter{#2}%
        \hyper@makecurrent{#2}%
        \let\Hy@dth@currentHref\@currentHref
        \def\@tempa{%
          \@oparg{\@begintheorem{#3}{\csname the#2\endcsname}}[]%
        }%
      \fi
      \@tempa
    }%

Summa summarum: **hyperref** replaces `\refstepcounter` by `\H@refstepcounter..\hyper@makecurrent`.

The effect is:

No anchors will be placed in the output-file while names of anchors are supplied to the `\label`-command. Thus referencing via `\ref` and the likes yields that the engine (pdflatex or the like) attempts to create a hyperlink to an anchor that does not exist.

**Comment by davidcarlisle** [comment]
Thanks for the report. In your tex.sx answer you simply undid the hyperref
patch which is a good workaround but I guess a full fix would make the
patch work.

I'll test later but have you tried modifying the

\ifx\Hy@temp\@thm


test so that it tests against the current amsthm definition?

David

**Comment by Ulrich-Diez** [comment]
I just did. As expected by me this does not solve the problem. You end in the `\if`-branch of the comparison. Here you have:

    \def\@thm#1#2#3{%
      \ifhmode
        \unskip\unskip\par
      \fi
      \normalfont
      \trivlist
      \let\thmheadnl\relax
      \let\thm@swap\@gobble
      \let\thm@indent\indent % indent
      \thm@headfont{\scshape}% heading font small caps
      \thm@notefont{\fontseries\mddefault\upshape}%
      \thm@headpunct{.}% add period after heading
      \thm@headsep 5\p@ plus\p@ minus\p@\relax
      \thm@space@setup
      #1% style overrides
      \@topsep \thm@preskip               % used by thm head
      \@topsepadd \thm@postskip           % used by \@endparenv
      \def\dth@counter{#2}%
      \ifx\@empty\dth@counter
        \def\@tempa{%
          \@oparg{\@begintheorem{#3}{}}[]%
        }%
      \else
        \H@refstepcounter{#2}%
        \hyper@makecurrent{#2}%
        \let\Hy@dth@currentHref\@currentHref
        \def\@tempa{%
          \@oparg{\@begintheorem{#3}{\csname the#2\endcsname}}[]%
        }%
      \fi
      \@tempa
    }%

In that definition too is nothing that does place an anchor for hyperlinks while there are things that do provide anchor-names for the `\label`-macro.

(If I got it right:

`\H@refstepcounter` is the LaTeX2e-Kernel variant which does increment a counter by 1 but does not place an anchor.

`\hyper@makecurrent` does create an anchor name but does not place an anchor.

`\let\Hy@dth@currentHref\@currentHref` does provide that anchor-name as `\@currentlabel` to the `\label`-macro.  )

**Comment by davidcarlisle** [comment]
thanks for looking, I'm a bit distracted at present but will try to sort this out soon.

**Comment by Ulrich-Diez** [comment]
I guess with more recent releases of **amsthm** patching is not needed at all. ;-)
If I got right, the old comments in hyperref.dtx/hyperref.pdf by Sebastian Rahtz, in the past there were problems about **ams** using low-level-commands instead of LaTeX2e-commands like `\refstepcounter` etc. Nowadays **amsthm** does use `\refstepcounter` etc.

**Comment by Ulrich-Diez** [comment]
The issue can be closed/deleted. 

I made a big big mistake:

I overlooked that in the mentioned TeX LaTeX StackExchange-Issue the theorem-like environment does nest its' labels into the optional argument of an \item of a \list. Thus in this case the box holding the label, `\@labels`, gets typeset by the LaTeX2e-kernel-command `\@item` and not by whatsoever command of one of the packages  **amsthm**/**hyperref**.

Thus there is no issue with these packages, but it's up to the user to pass - besides the content of the label - also the directive to place a hyperlink-anchor.

Sorry for the noise. Sorry for any inconvenience.
